rcl_action/src/rcl_action/action_server.c

```diff
@@ -0,0 +1,644 @@
+    &action_server->impl->Type ## _service, \
+    node, \
+    type_support->Type ## _service_type_support, \
+    Type ## _service_name, \
+    &Type ## _service_options); \
+  allocator.deallocate(Type ## _service_name, allocator.state); \
+  if (RCL_RET_OK != ret) { \
+    if (RCL_RET_BAD_ALLOC == ret) { \
+      ret = RCL_RET_BAD_ALLOC; \
+    } else { \
+      ret = RCL_RET_ERROR; \
+    } \
+    goto fail; \
+  }
+
+#define PUBLISHER_INIT(Type) \
+  char * Type ## _topic_name = NULL; \
+  ret = rcl_action_get_ ## Type ## _topic_name(action_name, allocator, &Type ## _topic_name); \
+  if (RCL_RET_OK != ret) { \
+    RCL_SET_ERROR_MSG("failed to get " #Type " topic name"); \
+    if (RCL_RET_BAD_ALLOC == ret) { \
+      ret = RCL_RET_BAD_ALLOC; \
+    } else { \
+      ret = RCL_RET_ERROR; \
+    } \
+    goto fail; \
+  } \
+  rcl_publisher_options_t Type ## _publisher_options = { \
+    .qos = options->Type ## _topic_qos, .allocator = allocator \
+  }; \
+  action_server->impl->Type ## _publisher = rcl_get_zero_initialized_publisher(); \
+  ret = rcl_publisher_init( \
+    &action_server->impl->Type ## _publisher, \
+    node, \
+    type_support->Type ## _message_type_support, \
+    Type ## _topic_name, \
+    &Type ## _publisher_options); \
+  allocator.deallocate(Type ## _topic_name, allocator.state); \
+  if (RCL_RET_OK != ret) { \
+    if (RCL_RET_BAD_ALLOC == ret) { \
+      ret = RCL_RET_BAD_ALLOC; \
+    } else { \
+      ret = RCL_RET_ERROR; \
+    } \
+    goto fail; \
+  }
+
+rcl_ret_t
+rcl_action_server_init(
+  rcl_action_server_t * action_server,
+  rcl_node_t * node,
+  const rosidl_action_type_support_t * type_support,
+  const char * action_name,
+  const rcl_action_server_options_t * options)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(action_server, RCL_RET_INVALID_ARGUMENT);
+  if (!rcl_node_is_valid(node)) {
+    return RCL_RET_NODE_INVALID;  // error already set
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(type_support, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(action_name, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(options, RCL_RET_INVALID_ARGUMENT);
+  rcl_allocator_t allocator = options->allocator;
+  RCL_CHECK_ALLOCATOR_WITH_MSG(&allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+
+  RCUTILS_LOG_DEBUG_NAMED(
+    ROS_PACKAGE_NAME, "Initializing action server for action name '%s'", action_name);
+  if (action_server->impl) {
+    RCL_SET_ERROR_MSG("action server already initialized, or memory was unintialized");
+    return RCL_RET_ALREADY_INIT;
+  }
+
+  // Allocate for action server
+  action_server->impl = (rcl_action_server_impl_t *)allocator.allocate(
+    sizeof(rcl_action_server_impl_t), allocator.state);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    action_server->impl, "allocating memory failed", return RCL_RET_BAD_ALLOC);
+
+  // TODO(jacobperron): Expand the given action name
+  // Test if name is valid
+
+  // Initialize services
+  rcl_ret_t ret;
+  SERVICE_INIT(goal);
```

Comment:
I agree that it is better to be consistent. I went with macros because it made for more concise code, cutting down on code duplication, but I could be convinced to use `static` functions instead.

Reply:
Not arguing in favor or against. Either way is fine for me as long as we use just one.

Reply:
I'm currently preferring the macro route because of its conciseness, if you don't mind updating #319 for consistency.